Sanitize email input when multiple attribute is set

lukewarlow opened the pull request:
#### 6ae622978d28c12e50253fab9c4cfc08bcffff1c
<pre>
Sanitize email input when multiple attribute is set
<a href="https://bugs.webkit.org/show_bug.cgi?id=256137">https://bugs.webkit.org/show_bug.cgi?id=256137</a>

Reviewed by Tim Nguyen and Chris Dumez.

The sanitizeValue function is now called when the multiple attribute is
changed. As per the spec <a href="https://html.spec.whatwg.org/#email-state-(type=email)">https://html.spec.whatwg.org/#email-state-(type=email)</a>

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/email-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/email.html:
* Source/WebCore/html/EmailInputType.cpp:
(WebCore::EmailInputType::attributeChanged):
* Source/WebCore/html/EmailInputType.h:

Canonical link: <a href="https://commits.webkit.org/263555@main">https://commits.webkit.org/263555@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c2aa2af7e2d240acc92af6693fd8ec0e773e1d31

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5050 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5180 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5357 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6570 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5140 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5043 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5374 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5155 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/5389 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5136 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5219 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4531 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6586 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2706 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4529 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/9542 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4569 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4602 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6209 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/4654 "Build is in progress. Recent messages:") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5015 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4114 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4501 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/4525 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1213 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8578 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4867 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/html/EmailInputType.cpp:
```diff
@@ -88,6 +91,14 @@ bool EmailInputType::supportsSelectionAPI() const
+    if (name == multipleAttr)
+        element()->setValueFromRenderer(sanitizeValue(element()->value()));
+
+    InputType::attributeChanged(name);
```
This should call the base class's implementation, which would be BaseTextInputType, not InputType

lukewarlow replied:
Apologies for that, fixed now.

cdumez reviewed (changes requested):
Note that there is a test failure.

pxlcoder commented on Source/WebCore/html/EmailInputType.cpp:
```diff
@@ -88,6 +91,14 @@ bool EmailInputType::supportsSelectionAPI() const
 
+void EmailInputType::attributeChanged(const QualifiedName& name)
+{
+    if (name == multipleAttr)
```
@cdumez Should we prefer `switch (name.nodeName())` now?

cdumez replied:
I don't think using a switch for a single check is better. My rule of thumb has been to use it when there are at least 3 checks (no perf benefit with less checks and I personally don't think it improves code readability much either).

pxlcoder reviewed (commented):
Please add a link to the [spec](https://html.spec.whatwg.org/#email-state-(type=email)) in the commit message.

cdumez reviewed (approved):
LGTM

lukewarlow committed:
Sanitize email input when multiple attribute is set
https://bugs.webkit.org/show_bug.cgi?id=256137

Reviewed by Tim Nguyen and Chris Dumez.

The sanitizeValue function is now called when the multiple attribute is
changed. As per the spec https://html.spec.whatwg.org/#email-state-(type=email)

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/email-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/email.html:
* Source/WebCore/html/EmailInputType.cpp:
(WebCore::EmailInputType::attributeChanged):
* Source/WebCore/html/EmailInputType.h:

Canonical link: https://commits.webkit.org/263555@main
